### PR TITLE
add test for caching, dissect moving parts

### PR DESCRIPTION
**Pull request by gutschilla**
Hi @kpanic ,

I couldn't help myself :grimacing: but had to try that out. Pretty darn cool thing! Would be awesome if you at some point could change `IO.puts` or `IO.write` to some string to append to.

Anyways, if you don't mind too much, I rearranged a few things here and there to break functionality apart so they can be tested. Which I didn't :-)

BTW: The "fix stupidity" commit refers to _my_ stupidity.

Thanks for your kind consideration, 
Martin

### CODE REVIEW

File: lib/forecastr.ex [hunk 1]
```diff
@@ -17,26 +17,58 @@ defmodule Forecastr do
   Forecastr.forecast(:in_five_days, "Berlin", %{units: :imperial})
   """
 
-  @spec forecast(atom(), String.t(), list()) :: {:ok, map()} | {:error, atom()}
+  @type when_to_forecast :: :today | :in_five_days
+  @type response :: map()
```
Comment (kpanic) [inline lib/forecastr.ex]: response is unused

File: lib/forecastr.ex [hunk 2]
```diff
@@ -17,26 +17,58 @@ defmodule Forecastr do
...
+  @spec forecast(when_to_forecast, String.t(), list()) :: {:ok, map()} | {:error, atom()}
   def forecast(when_to_forecast, query, params \\ %{units: :metric}) do
+    location = query |> String.downcase
```
Comment (kpanic) [inline lib/forecastr.ex]: Parens :stuck_out_tongue_winking_eye: `String.downcase()`
`String.downcase(query)`

File: lib/forecastr.ex [hunk 3]
```diff
@@ -17,26 +17,58 @@ defmodule Forecastr do
...
+    end
+  end
 
-        with {:ok, response} <- backend.weather(when_to_forecast, query, params),
-             expiration_minutes = Application.fetch_env!(:forecastr, :ttl) do
-          :ok = Forecastr.Cache.set(query, response, ttl: expiration_minutes)
+  # fetch form cache doesn't take when_to_forecast into account?
+  @spec fetch_from_cache(query) :: :ok
+  def fetch_from_cache(query) do
+    case query |> Forecastr.Cache.get do
```
Comment (kpanic) [inline lib/forecastr.ex]: `Forecastr.Cache.get(query)`

File: lib/forecastr.ex [hunk 4]
```diff
@@ -17,26 +17,58 @@ defmodule Forecastr do
...
-    query = String.downcase(query)
+    response |> renderer.render()
+  end
 
-    # TODO: distinguish between :today and :in_five_days cache
-    case Forecastr.Cache.get(query) do
-      nil ->
-        backend = Application.get_env(:forecastr, :backend)
+  @type query :: String.t()
+  @spec perform_query(query, when_to_forecast, map()) :: String.t()
```
Comment (kpanic) [inline lib/forecastr.ex]: `String.t()` shouldn't be:

` {:ok, map()} | {:error, atom()}`

File: lib/forecastr.ex [hunk 5]
```diff
@@ -17,26 +17,58 @@ defmodule Forecastr do
...
+    else
+      {:get_cache, {:ok, response}} -> {:ok, response}
+      {:fetch, _}                   -> {:error, :fetch_from_backend_failed}
+    end
+  end
 
-        with {:ok, response} <- backend.weather(when_to_forecast, query, params),
-             expiration_minutes = Application.fetch_env!(:forecastr, :ttl) do
-          :ok = Forecastr.Cache.set(query, response, ttl: expiration_minutes)
+  # fetch form cache doesn't take when_to_forecast into account?
```
Comment (kpanic) [inline lib/forecastr.ex]: Currently no, see TODO note here https://github.com/kpanic/forecastr/pull/1/files#diff-5f151367a5e3a0c573f5f36e29c84d21L25 :point_up:

File: lib/forecastr.ex [hunk 6]
```diff
@@ -17,26 +17,58 @@ defmodule Forecastr do
...
+    backend.weather(when_to_forecast, query, params)
+  end
 
-      response ->
-        response
-    end
-    |> renderer.render()
+  @spec cache_response(query, response) :: :ok | atom()
+  def cache_response(query, response) do
+    expiration_minutes = Application.get_env(:forecastr, :ttl, 10 * 60_000)
```
Comment (kpanic) [inline lib/forecastr.ex:69]: :+1: for the default argument

File: lib/forecastr.ex [hunk 7]
```diff
@@ -17,26 +17,58 @@ defmodule Forecastr do
   Forecastr.forecast(:in_five_days, "Berlin", %{units: :imperial})
   """
 
-  @spec forecast(atom(), String.t(), list()) :: {:ok, map()} | {:error, atom()}
+  @type when_to_forecast :: :today | :in_five_days
+  @type response :: map()
+  @spec forecast(when_to_forecast, String.t(), list()) :: {:ok, map()} | {:error, atom()}
```
Comment (kpanic) [inline lib/forecastr.ex]: Shouldn't be `:: binary() | {:error, atom()}`? It will be a string returned

**Review by kpanic — CHANGES_REQUESTED** [review]
Hey @gutschilla ! First of all, thanks for the first PR! :1st_place_medal: 
I left some minor comments :wink: 
much :heart: :green_heart: :yellow_heart: 

Thank you! for your contribution :bow:

**Comment by kpanic** [comment]
@gutschilla regarding `IO.write` `IO.puts` I would like to keep them in the renderer, because at some point I would like to add the possibility to dynamically select the renderer (think about a json/html/png/etc renderer!)

btw typespecs! Yey! I was too lazy to add them :stuck_out_tongue_winking_eye:

File: lib/forecastr.ex (hunk 7 above)
Comment (gutschilla) [reply lib/forecastr.ex]: As far as I cann see `:ok` will be returned. The strings themselves get IO'ed to STDOUT

File: lib/forecastr.ex (hunk 4 above)
Comment (gutschilla) [reply lib/forecastr.ex]: Very true!

File: lib/forecastr.ex (hunk 1 above)
Comment (gutschilla) [reply lib/forecastr.ex]: Ah, now it is

File: lib/forecastr.ex (hunk 2 above)
Comment (gutschilla) [reply lib/forecastr.ex]: A yeah, shouldn't ignore warnings :grimacing: – thanks for pointing me there

File: lib/forecastr.ex (hunk 3 above)
Comment (gutschilla) [reply lib/forecastr.ex]: It's still the Elixit 0.9-guy in me, I am sorry. Fixed this.

File: lib/forecastr.ex (hunk 6 above)
Comment (gutschilla) [reply lib/forecastr.ex:69]: I wonder, if this is duplicate now when I look at the config

**Comment by gutschilla** [comment]
@kpanic Thanks for your kind review. I addressed all the issues.

**Review by kpanic — APPROVED** [review]

**Pull request merged**